faiss/gpu/impl/IcmEncoder.cu

```diff
@@ -0,0 +1,468 @@
+    computeUnaryTerms(uterm.data(), x.data(), codebooks.data(), n);
+
+    DeviceTensor<int32_t, 2, true> bestCodes(
+            res.get(), makeTempAlloc(AllocType::Other, stream), {n, M});
+    fromDevice<int32_t, 2>(codes, bestCodes.data(), stream);
+
+    DeviceTensor<float, 1, true> bestObjs(
+            res.get(), makeTempAlloc(AllocType::Other, stream), {n});
+
+    DeviceTensor<float, 1, true> objs(
+            res.get(), makeTempAlloc(AllocType::Other, stream), {n});
+
+    // compute how much shared memory we need
+    const int evaluateSmem = sizeof(float) * (dims + kWarpSize - 1) / kWarpSize;
+    const int encodeSmem =
+            sizeof(Pair<float, int>) * (K + kWarpSize - 1) / kWarpSize;
+
+    // compute the reconstruction error for each vector
+    runEvaluation<M><<<n, dims, evaluateSmem, stream>>>(
+            x.data(),
+            codebooks.data(),
+            codes.data(),
+            bestObjs.data(),
+            n,
+            K,
+            dims);
+
+    int blockSize = 256;
+    int numBlocks = (n + blockSize - 1) / blockSize;
+
+    for (int i = 0; i < ilsIters; i++) {
+        runCodesPerturbation<M><<<numBlocks, blockSize, 0, stream>>>(
+                gen(), codes.data(), n, K, nperts);
+
+        // perform icm encoding
+        for (int j = 0; j < icmIters; j++) {
+            for (int m = 0; m < M; m++) {
+                runIcmEncodeStep<M><<<n, K, encodeSmem, stream>>>(
+                        uterm[m].data(), bterm[m].data(), codes.data(), K, m);
+            }
+        }
+
+        // compute the reconstruction error for each vector given codes
+        runEvaluation<M><<<n, dims, evaluateSmem, stream>>>(
+                x.data(),
+                codebooks.data(),
+                codes.data(),
+                objs.data(),
+                n,
+                K,
+                dims);
+
+        // if objs[i] < best_objs[i], replace best_codes[i] with codes[i]
+        runCodesSelection<M><<<numBlocks, blockSize, 0, stream>>>(
+                bestCodes.data(),
+                bestObjs.data(),
+                codes.data(),
+                objs.data(),
+                n);
+
+        codes.copyFrom(bestCodes, stream);
+    }
+
+    // copy back to host memory
+    fromDevice<int32_t, 2>(bestCodes, codesHost, stream);
+}
+
+void IcmEncoderImpl::encode(
+        int32_t* codes,
+        const float* x,
+        const float* codebooks,
+        std::mt19937& gen,
+        int n,
+        int nperts,
+        int ilsIters,
+        int icmIters) const {
+    FAISS_THROW_IF_NOT(M <= 64 || M == 128);
+
+#define DISPATCH_M(m)                                                        \
+    case m:                                                                  \
+        encodeImpl<m>(                                                       \
+                codes, x, codebooks, gen, n, K, nperts, ilsIters, icmIters); \
+        break
+
+    switch (M) {
+        DISPATCH_M(1);
```

Comment:
Template instantiations are the main cause of code bloat in cuda faiss. 
Please either make a generic version that can run with non-fixed M or select a subset of M values: 
for example 1 to 10 + powers of 2 + 3 * powers of 2.